chore(readme): add an example slash command (#656)
nextcord/nextcord: README.rst

* python-dev (e.g. ``python3.8-dev`` for Python 3.8)

Quick Example (with commands.Bot)

> ~~~~~~~~~~~~~

Reviewer: This line needs to be extended to the same length as the heading.

But I don't think it is necessary to include "(with commands.Bot)"

Reviewer 2: I think I added ooliver's changes.

Reviewer 3: but what is the need to explain to new users what class an example is using

Reviewer 2: Just to help them with differences and compare both classes before they choose a class?

* python-dev (e.g. ``python3.8-dev`` for Python 3.8)

> Quick Example (with ``commands.Bot``)
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: I don't think the long heading is necessary as you can see it in the example itself
```suggestion
Quick Example
~~~~~~~~~~~~~
```

    from nextcord.ext import commands

    bot = commands.Bot()

    @bot.slash_command(description="Replies with pong!")
    async def ping(interaction: nextcord.Interaction):
        await interaction.send("Pong!", ephemeral=True)

>         

Reviewer 3: there seems to be trailing whitespace on this line, could you remove the unnecessar spacing like shown in pre-commit?

Author: Check now please.

Reviewer 3: I'm being serious, what are you doing? Just have a look at the diff here, you see you added 2 extra not-needed newlines, remove those, not anything else.

Author: There you go.

Reviewer 3: You'll want to read my last sentence and the failing check.

Author: Pre-commit should have fixed it by now. Sorry, didn't understand what you meant clearly.

    from nextcord.ext import commands

    bot = commands.Bot()

    @bot.slash_command(description="Replies with pong!")
    async def ping(interaction: nextcord.Interaction):
        await interaction.send("Pong!", ephemeral=True)

>     bot.run("token")
> 

Reviewer 3: ```suggestion
    bot.run("token")
```

> 

Reviewer 3: I know this is really picky but its annoying me, could you remove this extra newline?

Author: Done!

Reviewer 3: i meant the other way round, leave the newline before `bot.run` but remove the added newline after it.

Author: Check now.

>     bot.run("token")
> 
> 

Reviewer 3: ```suggestion
    bot.run("token")

```

    from nextcord.ext import commands

    bot = commands.Bot()

    @bot.slash_command(description="Replies with pong!")
    async def ping(interaction: nextcord.Interaction):
        await interaction.send("Pong!", ephemeral=True)
        
    bot.run("token")

>     

Reviewer 3: Confused as to why there is trailing whitespace now

Author: Is it better now?

Reviewer 3: you just removed another line, add that back
![image](https://user-images.githubusercontent.com/34910574/180519984-bdbc0ae0-e8d3-4da6-9a7e-53f250a5f240.png)

    from nextcord.ext import commands

    bot = commands.Bot()

    @bot.slash_command(description="Replies with pong!")
    async def ping(interaction: nextcord.Interaction):
        await interaction.send("Pong!", ephemeral=True)

    bot.run("token")

> It is recommended that you use ``commands.Bot`` if you wish to use prefix commands, as although both ``nextcord.Client`` and ``commands.Bot`` allow you to interface with Discord, ``commands.Bot`` provides an easier way to make prefix commands with a framework. Make sure to not use both versions, though, as only one of them will actually be connected to Discord.

Reviewer 3: I'm not too sure how I feel about this line overall. It just throws information onto someone who has only just seen the library (otherwise why would they read the example) which they do not need from the README. The docs can be used instead to demonstrate this.

Reviewer: I agree with this, just having a simple example is enough. (It also can be confusing since it says commands.Bot is recommended for prefix commands, but the example uses commands.Bot for slash. This is fine, but can be confusing to read them together.)

Author: > I agree with this, just having a simple example is enough. (It also can be confusing since it says commands.Bot is recommended for prefix commands, but the example uses commands.Bot for slash. This is fine, but can be confusing to read them together.)

I've updated the file.

Reviewer 3: Now it is just downright incorrect, `commands.Bot` provides no extra slash-specific additions.

Author: What do you mean? bot.slash_command uses commands.Bot

Reviewer 3: `Client.slash_command` also exists, your description on why `commands.Bot` is good is just unnecessarily complex. There is no need for it, users haven't even used the library yet if they are reading the example.

Reviewer: I think it's probably best to just remove the paragraph and maybe add something in the main documentation in another PR

Author: > I think it's probably best to just remove the paragraph and maybe add something in the main documentation in another PR

Did that just now.

    from nextcord.ext import commands

    bot = commands.Bot()

    @bot.slash_command(description="Replies with pong!")
    async def ping(interaction: nextcord.Interaction):
        await interaction.send("Pong!", ephemeral=True)
        
    bot.run("token")

>     

Reviewer 3: ```suggestion

```

    bot = commands.Bot(command_prefix='$')

    @bot.slash_command(description="Replies with pong!")
    async def ping(interaction: nextcord.Interaction):
        await interaction.send("Pong!", ephemeral=True)

    bot.run("token")

> It is recommended that you use ``commands.Bot``, as although both ``nextcord.Client`` and ``commands.Bot`` allow you to interface with Discord, ``commands.Bot`` provides an easier way to make prefix commands with a framework. Make sure to not use both versions, though, as only one of them will actually be connected to Discord.

Reviewer 4: id specify that its recommended *if you wish to add prefix commands*?

Reviewer: I think it's still recommended due to Cogs, listeners, some checks?, error handling, among other things only existing with commands.Bot. It would be nice if it didn't have to all be connected, but that's how it is as of now I believe.

Author: Updated, please check now.

Reviewer: I think it was better before. It's kind of repetitive to say twice that its recommended.

If you're going to say its recommended for prefix commands (Epic's suggestion), that should go at the beginning of the sentence (eg. "It is recommended that you use ``commands.Bot`` if you wish to add prefix commands, ..."), although I think it's not even needed at all to specify and it was fine as it was before.

Just waiting on @TAG-Epic to see what they think it should say given my comment.

Author: Ok, will fix something real quick as I see some wording error.

> It is recommended that you use ``commands.Bot``, as although both ``nextcord.Client`` and ``commands.Bot`` allow you to interface with Discord, ``commands.Bot`` provides an easier way to make prefix commands with a framework and it is recommended to use ``commands.Bot`` for prefix commands & slash commands. Make sure to not use both versions, though, as only one of them will actually be connected to Discord.

Reviewer: The wording is unnecessarily long here.

In my opinion, it is not necessary at all to specify where it is recommended since commands.Bot currently allows features that are useful even if you are not using prefix commands.

```suggestion
It is recommended that you use ``commands.Bot``, as although both ``nextcord.Client`` and ``commands.Bot`` allow you to interface with Discord, ``commands.Bot`` provides an easier way to make prefix commands with a framework. Make sure to not use both versions, though, as only one of them will actually be connected to Discord.
```

With @TAG-Epic's suggestion taken into account, it could be simplified to:

```suggestion
It is recommended that you use ``commands.Bot`` if you wish to use prefix commands, as although both ``nextcord.Client`` and ``commands.Bot`` allow you to interface with Discord, ``commands.Bot`` provides an easier way to make prefix commands with a framework. Make sure to not use both versions, though, as only one of them will actually be connected to Discord.
```

@TAG-Epic's opinion is requested on this before applying one of the above changes.

Reviewer 4: Id add a bit of info about cogs & multiple listeners aswell with ext.commands, but other than that it looks very good!

    bot = commands.Bot(command_prefix='$')

    @bot.slash_command(description="Replies with pong!")
    async def ping(interaction: nextcord.Interaction):
        await interaction.send("Pong!", ephemeral=True)

>     bot.run('token')

Reviewer 3: ```suggestion
    bot.run("token")
```

> It is recommended that you use ``commands.Bot``, as although both ``nextcord.Client`` and ``commands.Bot`` allow you to interface with Discord, ``commands.Bot`` provides an easier way to make commands with a framework. Make sure to not use both versions, though, as only one of them will actually be connected to Discord.

Reviewer 3: ```suggestion
It is recommended that you use ``commands.Bot``, as although both ``nextcord.Client`` and ``commands.Bot`` allow you to interface with Discord, ``commands.Bot`` provides an easier way to make prefix commands with a framework. Make sure to not use both versions, though, as only one of them will actually be connected to Discord.
```
reduce ambiguity as there is a slash command example above

    bot = commands.Bot(command_prefix='$')

    @bot.slash_command(description="Replies with pong!")
    async def ping(interaction: nextcord.Interaction):

>         await interaction.send("Pong!")

Reviewer: I think having 1 example command would be enough. Maybe remove the "ephemeral" command, and instead keep just the "ping" command but make the "Pong!" response ephemeral.

    @bot.command()
    async def say(ctx, *, arg):
         await ctx.message.delete()

>          await ctx.send(f"{ctx.author.mention}'s message: {arg}")

Reviewer 3: either edit the quotes to be single here or double in the rest (latter preferred)

> 

Reviewer: I don't think we need a second example in the readme. That is what the `examples` folder and documentation are for.

Reviewer 2: Good idea, I will work on this.

>     async def epehemeral(interaction: nextcord.Interaction):

Reviewer: ```suggestion
    async def ephemeral(interaction: nextcord.Interaction):
```

>         await interaction.send("Ephemeral message!", ephemeral=True)

Reviewer 2: Outdated

    bot.run('token')

> It is recommended that you use commands.Bot version, as although both client (nextcord.Client) and commands.Bot have the same stuff, commands.Bot is easy to use rather than the client version of nextcord. Make sure to not use both versions though as it can can be confusing and conflicts with your code upon using both in your code.

Reviewer 3: ```suggestion
It is recommended that you use ``commands.Bot``, as although both client (``nextcord.Client``) and ``commands.Bot`` allow you to interface with discord, ``commands.Bot`` provides an easier way to make commands with a framework. Make sure to not use both versions though as only one of them will actually be connected to discord.
```

> It is recommended that you use commands.Bot version, as although both client (nextcord.Client) and commands.Bot have the same stuff, commands.Bot is easy to use rather than the client version of nextcord. Make sure to not use both versions though as it can can be confusing and conflicts with your code upon using both in your code.

Reviewer: This explanation hardly explains why `commands.Bot` is used over `nextcord.Client`.

`nextcord.Client` is fine for bots that don't need message-based commands, Cogs, listeners, or the built-in error handling. You are right that `commands.Bot` makes certain things easier than `nextcord.Client` and that you should not use both. I don't think it is necessary to explain that here though.

> It is recommended that you use ``commands.Bot``, as although both client (``nextcord.Client``) and ``commands.Bot`` allow you to interface with discord, ``commands.Bot`` provides an easier way to make commands with a framework. Make sure to not use both versions though as only one of them will actually be connected to discord.

Reviewer: ```suggestion
It is recommended that you use ``commands.Bot``, as although both ``nextcord.Client`` and ``commands.Bot`` allow you to interface with Discord, ``commands.Bot`` provides an easier way to make commands with a framework. Make sure to not use both versions, though, as only one of them will actually be connected to Discord.
```

- `Documentation <https://docs.nextcord.dev/>`_
- `Official Discord Server <https://discord.gg/ZebatWssCB>`_
- `Discord API <https://discord.gg/discord-api>`_

> - `Nextcord's Github <https://www.github.com/nextcrd/nextcord>`_

Reviewer: This is spelled wrong, but also, what's the point of including the GitHub link in the GitHub readme? You're already on the page when you're looking at it.

If it's for Pypi, it's also not needed since there is already a "Homepage" and "Issue tracker" link on Pypi that links here.

Reviewer 2: Just for the nextcord readthedocs page.

Reviewer 3: thats a completely different file entirely

Reviewer 2: Removed, thanks
